### DISCUSSION TITLE
[Idea] Unify sorting options for all pages

### DISCUSSION THREAD

**Discussion by Hlavtox** [PrestaShop/PrestaShop#32248] [Ideas]
In current state, there is a huge mess in a way product sorting options are defined and configured.

**Context**
- Category without faceted search has some options
- Faceted search sets it's own
- Every controller like new products, best sellers etc. has it's own
- Then, you can configure default sort in Prestashop setting

**How it behaves**
1. I currently have set in my PS settings my default sort to Quantity.
2. I open category with faceted search.
3. Products are sorted by Quantity, but in the sort options list there is only price asc, price desc, name asc, name desc and position. And position is selected.

**Proposed solution**
Merchant should be able to set up default sorting for each controller and for faceted search module.

#### Additional information
* PrestaShop version: 1.7.6.2
* PHP version: 7.2

[a comment by an automated account is not shown]

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

Thanks for your report.
Is it your issue similar to this: https://github.com/PrestaShop/PrestaShop/issues/14210

>, but in the sort options list there is only price asc, price desc, name asc, name desc and position. And position is selected.

With PS1.7.6.7., I have an empty box selected
![image](https://user-images.githubusercontent.com/16067358/89892120-7c1d4a00-dbd6-11ea-8811-3bd1cd0990c7.png)

![image](https://user-images.githubusercontent.com/16067358/89892100-7162b500-dbd6-11ea-9ae8-c6c37d1aa7b7.png)


Thanks to check and feedback.

**Comment by Hlavtox** [comment]
@khouloudbelguith The white box is a matter of template, ok.

The bigger question is still in, there is a mess in it, I think.

There are sort ways in BO that are not available in FO anywhere and so on.

**Comment by khouloudbelguith** [comment]
@Hlavtox, thanks for your feedback.
Ping @PrestaShop/prestashop-product-team what do you think?
Can we add the missing order methods in the FO?
![image](https://user-images.githubusercontent.com/16067358/89991569-4fc00700-dc84-11ea-84fe-8e095759c0a0.png)

Thanks!

**Comment by marionf** [comment]
Hello @Hlavtox 

The default order option selected in the BO corresponds to the "relevance" option in FO (selected by default when you opens a category in FO)
So, if you select for exemple "Product add date" in your BO, your products in categories in FO will be sorted this way when "relevance" is selected. 
But your visitors can choose to sort products another way (by name or price)
Product date, brand, product quantity and product reference are not displayed in FO as it makes no sense for a visitor

**Comment by Hlavtox** [comment]
@marionf Thats not true! Relevance is position inside of category. The sort you get when loading the page is the one from BO.

First image is when loading page (product quantity order) and the second one is relevance (product position order).

![D0D98CA1-5F66-4AC4-9671-E9C28D9C7C1D](https://user-images.githubusercontent.com/6097524/89995077-513ffe00-dc89-11ea-82dc-dca81b81f9b6.png)

![8251774F-E999-48BB-941D-4812F64961A1](https://user-images.githubusercontent.com/6097524/89995097-57ce7580-dc89-11ea-9880-861995ea2e86.png)

**Comment by NeOMakinG** [comment]
@marionf you sure that product add date and product brand doesn't make sense at all in FO ?

I mean, as an user, I could be interested about sorting products by brands or by added date in order to see recents products or even sorter by brands to search for a specific brand or juste sort it !

There is another problem on the "added date", we never show the added date on product pages, sorting by it would suppose that we show the date somewhere in FO, else it would be a bit blurry

**Comment by marionf** [comment]
> @marionf Thats not true! Relevance is position inside of category. The sort you get when loading the page is the one from BO.

You're right, sorry I missed that

> I could be interested about sorting products by brands or by added date in order to see recents products or even sorter by brands to search for a specific brand or juste sort it !

About the brand, this is possible to search by brand through faceted search module

> There is another problem on the "added date", we never show the added date on product pages, sorting by it would suppose that we show the date somewhere in FO, else it would be a bit blurry

Unless if you call this option "News"

Anyway, I think there is a study to do on this topic !

**Comment by Hlavtox** [comment]
@marionf @NeOMakinG 

Another thing. @dgonzalez360 added a sorting by sales to faceted search https://github.com/PrestaShop/ps_facetedsearch/pull/172, but this order is missing from BO list of sort orders. What if an user wants to select it by default? He cannot. We need to unify it and make it configurable.

**Comment by Hlavtox** [comment]
Ping @matks @eternoendless @marionf 

What if we do it like this? Then we add a selectbox to choose the default one, with ASC/DESC setting.

![sorting](https://user-images.githubusercontent.com/6097524/90030034-db08bf00-dcbb-11ea-87c3-55d05cd6874c.JPG)

**Comment by futurowebpl** [comment]
> @marionf @NeOMakinG
> 
> Another thing. @dgonzalez360 added a sorting by sales to faceted search [PrestaShop/ps_facetedsearch#172](https://github.com/PrestaShop/ps_facetedsearch/pull/172), but this order is missing from BO list of sort orders. What if an user wants to select it by default? He cannot. We need to unify it and make it configurable.

Absolutely yes, this is a critical thread.

 @dgonzalez360 How can We enforce the default sort by quantity sold, since there is no such item in Preferences -> Products?

**Comment by dgonzalez360** [comment]
Hi @futurowebpl, 

I'm not sure about the best way to achieve it. Add a default selector for sorting on the ps_facetedsearch is easy but I think is a bit confusing for Prestashop users because you can select two different options (One inside Preferences -> Products and other inside Faceted Search Module). 

The bigger issue here is that we can't unify the sorting order for product list controllers, each module can add their own sorting options and custom options are not available for all modules. 

I think one of the ways to achieve this is to create new hooks (One for the productSortOrderSelection and one for the getProductsOrder inside Tools.php), then we can use the default Prestashop Sort Options plus all the options available inside modules.

With two hooks we can extend the sorting order options for all modules and make it available for all controllers unifying the layout. 

What do you guys think? @marionf @NeOMakinG @Hlavtox

**Comment by futurowebpl** [comment]
Unless there's an easier way

From what I looked, unfortunately it involves changing a lot of files from the core - hence maybe some simpler fix?

Default sorting after most purchased - I am sure that everyone will use this sort of sort, all others are very rarely used.

**Comment by futurowebpl** [comment]
@khouloudbelguith the inability to force this sort by default is a waste of a very relevant functionality that has been missing for a long time,

Can anyone say whether it will be done or the developers should look for other ways?

**Comment by khouloudbelguith** [comment]
Ping @PrestaShop/product-team friendly reminder :slightly_smiling_face: 

Thanks!

**Comment by futurowebpl** [comment]
@khouloudbelguith any news for this task?

**Comment by kpodemski** [comment]
Hey @futurowebpl 

it's quite a complex topic and I'm not sure if we can deliver this before 1.7.7 release but let me assure you that it's something my customers and I are struggling with too, I'll look into this topic personally in the near future

**Comment by futurowebpl** [comment]
@kpodemski have you looked at this topic? :)

**Comment by tom-combet** [comment]
Something that makes no sense too: default sort order when searching for products is [this one](https://github.com/PrestaShop/PrestaShop/blob/977b59b8daefca5fe97c2d9bc7a170684ad2c206/controllers/front/listing/SearchController.php#L97) but it is absent from default [available sort orders](https://github.com/PrestaShop/PrestaShop/blob/977b59b8daefca5fe97c2d9bc7a170684ad2c206/src/Core/Product/Search/SortOrdersCollection.php#L54-L67)...

Shouldn't the method getDefaults return every possible sort orders options? At least, the ones from the back-office would indeed makes sense.
